ndk-build/src/cargo.rs

```diff
@@ -14,12 +14,43 @@ pub fn cargo_ndk(
     let clang_target = format!("--target={}{}", target.ndk_llvm_triple(), sdk_version);
     let mut cargo = Command::new("cargo");
 
-    // Read initial RUSTFLAGS
+    const SEP: &str = "\x1f";
+
+    // Read initial CARGO_ENCODED_/RUSTFLAGS
     let mut rustflags = match std::env::var("CARGO_ENCODED_RUSTFLAGS") {
-        Ok(val) => val,
-        Err(std::env::VarError::NotPresent) => "".to_string(),
+        Ok(val) => {
+            if std::env::var_os("RUSTFLAGS").is_some() {
+                panic!(
+                    "Both `CARGO_ENCODED_RUSTFLAGS` and `RUSTFLAGS` were found in the environment, please clear one or the other before invoking this script"
+                );
+            }
+
+            val
+        }
+        Err(std::env::VarError::NotPresent) => {
+            match std::env::var("RUSTFLAGS") {
+                Ok(val) => {
+                    // Note cargo also does this very simplified parsing of only
+                    // splitting on spaces, ignoring things like quoting
+                    // https://github.com/rust-lang/cargo/blob/f6de921a5d807746e972d9d10a4d8e1ca21e1b1f/src/cargo/core/compiler/build_context/target_info.rs#L682-L690
+                    let to_join: Vec<_> = val
+                        .split(' ')
+                        .map(|s| s.trim())
+                        .filter(|s| !s.is_empty())
+                        .collect();
+
+                    cargo.env_remove("RUSTFLAGS");
+
+                    to_join.join(SEP)
```

Comment:
Feel free to swap these and get rid of `to_join`, you already have the value of `RUSTFLAGS` so it can be deleted from the (nested, anyway) `cargo` env right after `Ok(val)`.